src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs

```diff
@@ -0,0 +1,422 @@
+using Microsoft.Azure.WebJobs.Script.Binding;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Scripting;
+using Microsoft.CodeAnalysis.Scripting;
+
+namespace Microsoft.Azure.WebJobs.Script.Description
+{
+    public class CSharpFunctionInvoker : ScriptFunctionInvokerBase
+    {
+        private const string ScriptClassName = "Submission#0";
+
+        private readonly ScriptHost _host;
+        private readonly DictionaryJsonConverter _dictionaryJsonConverter = new DictionaryJsonConverter();
+        private readonly TraceWriter _fileTraceWriter;
+        private readonly FunctionMetadata _functionMetadata;
+        private readonly BindingMetadata _trigger;
+        private readonly Collection<FunctionBinding> _inputBindings;
+        private readonly Collection<FunctionBinding> _outputBindings;
+        private readonly FileSystemWatcher _fileWatcher;
+        private readonly string _triggerParameterName;
+        private readonly bool _omitInputParameter;
+        private readonly IFunctionEntryPointResolver _functionEntryPointResolver;
+        private MethodInfo _function;
+        private FunctionMetadataResolver _metadataResolver;
+        private readonly FunctionAssemblyLoader _assemblyLoader;
+        private Action _reloadScript;
+        private Action _restorePackages;
+
+        private static readonly string[] _watchedFileTypes = { ".cs", ".csx", ".dll", ".exe" };      
+
+        public CSharpFunctionInvoker(ScriptHost host, BindingMetadata trigger, FunctionMetadata metadata, bool omitInputParameter,
+            Collection<FunctionBinding> inputBindings, Collection<FunctionBinding> outputBindings, 
+            IFunctionEntryPointResolver functionEntryPointResolver, FunctionAssemblyLoader assemblyLoader)
+        {
+            _host = host;
+            _trigger = trigger;
+            _triggerParameterName = trigger.Name;
+            _omitInputParameter = omitInputParameter;
+            _inputBindings = inputBindings;
+            _outputBindings = outputBindings;
+            _functionMetadata = metadata;
+            _functionEntryPointResolver = functionEntryPointResolver;
+            _fileWatcher = InitializeFileWatcher(host);
+            _assemblyLoader = assemblyLoader;
+
+            if (_host.ScriptConfig.FileLoggingEnabled)
+            {
+                string logFilePath = Path.Combine(_host.ScriptConfig.RootLogPath, "Function", _functionMetadata.Name);
+                _fileTraceWriter = new FileTraceWriter(logFilePath, TraceLevel.Verbose);
+            }
+            else
+            {
+                _fileTraceWriter = NullTraceWriter.Instance;
+            }
+
+            _metadataResolver = new FunctionMetadataResolver(metadata, _fileTraceWriter);
+
+            _reloadScript = Debounce(ReloadScript);
+            _restorePackages = Debounce(RestorePackages);
+        }
+
+        private Action Debounce(Action targetAction, int milliseconds = 500)
```

Comment:
Wondering if this should also be moved to somewhere where it can be a common helper. You could stuff it as a static on ActionExtensions if you wanted.